### DISCUSSION TITLE
Github Actions Mask Expression

### DISCUSSION THREAD

**Discussion by ugam-paul** [community/community#26032] [Actions]
<p>Hi everyone. I ran into a problem recently regarding github actions and it’s logs.<br>
I need to run a cli command like how it’s available for aws. The parameters contain username and password.<br>
What I did was, I created a JSON file with the username and password parameters, encrypted the same using GPG and stored the passphrase on github secrets.<br>
In my workflow, I decrypted the same and to parse it I used:<br>
content=<code>cat $TMP/secret.json</code><br>
content="<span class="math">{content//'%'/'%25'}"
content="</span>{content//<span class="math">'\n'/'%0A'}"
content="</span>{content//$’\r’/’%0D’}"<br>
echo “::set-output name=json::$content”</p>
<p>To pass the username and password I used:<br>
<span class="math">{{ fromJson(steps.var.outputs.json).username }}
</span>{{ fromJson(steps.var.outputs.json).password }}</p>
<p>The problem is when this expression gets evaluated it shows up in logs. To try and avoid this I did:<br>
echo “::add-mask::<span class="math">{{ fromJson(steps.var.outputs.json).username }}
echo "::set-output name=secret-username::</span>{{ fromJson(steps.var.outputs.json).username }}”<br>
echo “::add-mask::<span class="math">{{ fromJson(steps.var.outputs.json).password }}"
echo "::set-output name=secret-password::</span>{{ fromJson(steps.var.outputs.json).password }}”</p>
<p>Now in the CLI command the username and password seems to be masked, but still in the add-mask command it’s shown. It seems like expressions are evaluated and printed as commands before add-mask can register them.</p>
<p>Is there any workaround regarding the same?<br>
My apologies regarding the long post.</p>
<p>Edit: I can’t seem to get the formatting right for the post. My apologies for that as well.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/ugam-paul">@ugam-paul</a>,</p>
<p>As a workaround to avoid the username and password are displayed in the logs before mask, you can manually add a multi-line secret that contains some strings or keywords you want to mask in the logs.<br>
In this way, the strings or keywords in the logs will be hidden as long as they match any line you set in this secret.<br>
For example:<br>
Set a secret like as this.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/0ed927b91d9784b84c8162b0b6afad7cdf586e7d.png" data-download-href="/uploads/short-url/27lW7k0rJ4ymxRQJ88xPX4HI1id.png?dl=1" title="add_secret" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181094880-97febd36-2f7e-4e2c-866d-33a88ca5a7f5.png" alt="add_secret" data-base62-sha1="27lW7k0rJ4ymxRQJ88xPX4HI1id" width="690" height="331" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/0/0ed927b91d9784b84c8162b0b6afad7cdf586e7d_2_690x331.png, https://aws1.discourse-cdn.com/github/original/2X/0/0ed927b91d9784b84c8162b0b6afad7cdf586e7d.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/0/0ed927b91d9784b84c8162b0b6afad7cdf586e7d.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/0ed927b91d9784b84c8162b0b6afad7cdf586e7d_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">add_secret</span><span class="informations">929×446 12.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Then in the workflow, map this secret as a global (workflow level) environment variable.</p>
<pre><code class="lang-yaml">. . .
env:
  SOME_SECRETS: ${{ secrets.SOME_SECRETS }}

jobs:
  . . .
</code></pre>

**Comment by ugam-paul** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a>. Thank you for your response.<br>
While this is a good suggestion, it won’t work in my case as the secret could change and has to be dynamic. The only problem with the method I’m using is that expression evaluations show up in logs even if it is supposed to be masked. Since I’m using masking and expression evaluation in the same line.</p>
<p>echo “::add-mask::{{ fromJson(steps.var.outputs.json).username }}"</p>
<p>Anything else I could try regarding this case?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/ugam-paul">@ugam-paul</a>,</p>
<p>How about mask the whole json content?<br>
I see you have this step,</p>
<blockquote>
<p>echo “::set-output name=json::$content”</p>
</blockquote>
<p>Maybe you can try to add this step to see if it work.</p>
<blockquote>
<p>echo “::add-mask::{{ fromJson(steps.var.outputs.json) }}”</p>
</blockquote>

**Comment by ugam-paul** [comment]
<p><a class="mention" href="/brightran">@brightran</a></p>
<p>Tried this way. While in the logs, github shows the line as<br>
<code>“::add-mask::Object”</code><br>
In the subsequent line the credentials still show up.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/ugam-paul">@ugam-paul</a> ,</p>
<p>The following workaround should work:</p>
<ul>
<li>
<p>Add the command lines you need to execute in a bash script file (e.g. <strong>mask.sh</strong>).</p>
</li>
<li>
<p>Execute the bash script in your workflow.</p>
<pre><code class="lang-yaml">- name: mask secret
  id: mask
  run: |
    chmod +x ./mask.sh
    ./mask.sh
</code></pre>
</li>
</ul>
<p>When “<strong>run</strong>” a bash script file, the command lines in the script file will not be printed in the logs.</p>
<p>Here is a example I tested on my side:</p>
<ul>
<li>
<p>JSON file: <a href="https://github.com/TestWorkflowsRML/TestOutputs/blob/dfee4d8736c97b73d5c6d5e14f7d1a2b4aa387a0/secret.json" rel="nofollow noopener">https://github.com/TestWorkflowsRML/TestOutputs/blob/dfee4d8736c97b73d5c6d5e14f7d1a2b4aa387a0/secret.json</a></p>
</li>
<li>
<p>Script file: <a href="https://github.com/TestWorkflowsRML/TestOutputs/blob/dfee4d8736c97b73d5c6d5e14f7d1a2b4aa387a0/mask.sh" rel="nofollow noopener">https://github.com/TestWorkflowsRML/TestOutputs/blob/dfee4d8736c97b73d5c6d5e14f7d1a2b4aa387a0/mask.sh</a></p>
</li>
<li>
<p>Workflow: <a href="https://github.com/TestWorkflowsRML/TestOutputs/runs/1050549872?check_suite_focus=true" rel="nofollow noopener">https://github.com/TestWorkflowsRML/TestOutputs/runs/1050549872?check_suite_focus=true</a></p>
</li>
</ul>

**Comment by ugam-paul** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a>!</p>
<p>This worked like a charm!<br>
Thank you so much for the quick replies and an apt solution.</p>

**Comment by BrightRan** [comment]
<p>You’re welcome, <a class="mention" href="/ugam-paul">@ugam-paul</a> .<br>
And glad that my suggestion is helpful to you.</p>
